crates/sui-core/src/execution_engine.rs

```diff
@@ -450,6 +504,131 @@ fn pay<S>(
     Ok(())
 }
 
+fn pay_sui<S>(
```

Comment:
I haven't read into the details of the impl, but I hardly believe it works without any other changes. For instance, the gas object will be mutated twice, will its version be incremented twice as well?
Also if you look at the check_gas function in transaction_input_checker, we check that the amount + gas is always sufficient for the balance. This variant must hold otherwise there will be assertion failures during gas deduction.

Reply:
the gas object was `update_contents_without_version_change` in the txn execution and counts on gas deduction to increment the version, so still increment by 1 here, will test in the e2e test.

check_gas, great catch, I will want to want to check that as well, doing it now.